### PR TITLE
Subgraph: Add liquidity metrics

### PR DESCRIPTION
**Pull request by PJColombo**
This PR solves issue #98 by tracking Honeyswap's  `Pair` contracts in order to get the liquidity metrics.

### CODE REVIEW

File: packages/subgraph/src/helpers/pair.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import { Pair as PairContract } from '../../generated/GardensTemplate/Pair'
+import { Pair as PairTemplate } from '../../generated/templates'
+import { loadOrCreateOrg, loadOrCreatePair, loadTokenData, ZERO_ADDRESS } from '.'
+import { Organization as OrganizationEntity } from '../../generated/schema'
+
+function populatePairDataFromContract(pairAddress: Address, org: OrganizationEntity | null): void {
+  const pairEntity = loadOrCreatePair(pairAddress)
+  const pair = PairContract.bind(pairAddress)
+
+  const token0Id = loadTokenData(pair.token0())
```
Comment (0xGabi) [inline packages/subgraph/src/helpers/pair.ts:14]: I think here we should always have a `token` entity created.

**Review by 0xGabi — COMMENTED** [review]
This is more of a general comment that I notice. I think the `loadEntityData` functions were handled a bit wierd. As I understand they should never return `null` as we are always creating an entity if one doesn't exist. As a consequence, we are considering a lot of `Entity | null` types in several places that end up convoluting the code.

File: packages/subgraph/src/helpers/pair.ts (hunk 1 above)
Comment (PJColombo) [reply packages/subgraph/src/helpers/pair.ts]: I'll close this since the `Pair` entity has been removed.

**Comment by PJColombo** [comment]
> This is more of a general comment that I notice. I think the `loadEntityData` functions were handled a bit wierd. As I understand they should never return `null` as we are always creating an entity if one doesn't exist. As a consequence, we are considering a lot of `Entity | null` types in several places that end up convoluting the code.

I think we can use `return entity!` in cases where we are sure we're returning a non-null entity sure so we can avoid using the `Entity | null` type.

File: packages/subgraph/src/helpers/conviction.ts [hunk 2]
```diff
@@ -43,10 +44,15 @@ export function loadConvictionConfig(orgAddress: Address, appAddress: Address):
   if (stakeTokenId) {
     convictionConfig.stakeToken = stakeToken.toHexString()
 
-    // Set token for org
+    // Set token for org and vice versa
```
Comment (0xGabi) [inline packages/subgraph/src/helpers/conviction.ts:47]: I think this is all right for now. Regardless I was considering if having the initialization of the token here was the best place.

File: packages/subgraph/src/helpers/index.ts [hunk 3]
```diff
@@ -10,7 +10,13 @@ import {
 } from '../../generated/schema'
 import { STATUS_ACTIVE, STATUS_ACTIVE_NUM } from '../statuses'
 
+export const ZERO_ADDRESS = Address.fromString('0x0000000000000000000000000000000000000000')
+
 /// /// Token Entity //////
+export function isOrgToken(token: TokenEntity | null): boolean {
```
Comment (0xGabi) [inline packages/subgraph/src/helpers/index.ts:16]: What you think of using a more descriptive name for the function `isOrgToken`? Took me a bit of time to understand what was going on.

File: packages/subgraph/src/helpers/pair.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+export function setUpHoneyLiquidity(templateAddress: Address, orgAddress: Address): void {
+  const org = loadOrCreateOrg(orgAddress)
+  const gardensTemplate = GardensTemplateContract.bind(templateAddress)
+  const honeyswapRouter = HoneyswapRouterContract.bind(gardensTemplate.honeyswapRouter())
+  const honeyswapFactory = HoneyswapFactoryContract.bind(honeyswapRouter.factory())
+  // Note we assume org entity has token field set
+  const honeyDaoTokenPairAddress = honeyswapFactory.getPair(gardensTemplate.honeyToken(), Address.fromString(org.token))
+
+  if (honeyDaoTokenPairAddress.equals(ZERO_ADDRESS)) {
```
Comment (0xGabi) [inline packages/subgraph/src/helpers/pair.ts:46]: Here we are considering the 1Hive organization-specific case? Won't we?

**Review by 0xGabi — CHANGES_REQUESTED** [review]
Well done! Left a few comments. And a suggestion.

File: packages/subgraph/src/helpers/index.ts (hunk 3 above)
Comment (PJColombo) [reply packages/subgraph/src/helpers/index.ts:16]: What do you think about `checkForOrganization()` or `checkForOrg()` ?

File: packages/subgraph/src/helpers/pair.ts (hunk 4 above)
Comment (PJColombo) [reply packages/subgraph/src/helpers/pair.ts:46]: Yes, the subgraph is only tracking organization pairs.

File: packages/subgraph/src/helpers/conviction.ts (hunk 2 above)
Comment (PJColombo) [reply packages/subgraph/src/helpers/conviction.ts:47]: On second thought, maybe we can set the token in the [`setUpHoneyLiquidity()`](https://github.com/PJColombo/gardens/blob/f63f8320afb47f719698f248f3d56d555208ad39/packages/subgraph/src/helpers/pair.ts#L38) function, where we have access to the dao address. What do you think ?

File: packages/subgraph/src/helpers/index.ts (hunk 3 above)
Comment (fabriziovigevani) [reply packages/subgraph/src/helpers/index.ts:16]: Perhaps something like `tokenHasOrg`?

File: packages/subgraph/src/helpers/conviction.ts (hunk 2 above)
Comment (fabriziovigevani) [reply packages/subgraph/src/helpers/conviction.ts:47]: Yeah i think this makes sense. And we can be rest assured the org will have the associated token already.

One important thing that we need to take into account is the fact that for BYOT gardens, the token that is going to be paired with HNY will be the `org.wrappableToken` so we should check first if the org has a `wrappableToken` associated and use that one. In the case it doesn't fallback to the `org.token`

File: packages/subgraph/src/mappings/GardensTemplate.ts [hunk 5]
```diff
@@ -17,10 +17,14 @@ export function handleDeployDao(event: DeployDaoEvent): void {
 }
 
 export function handleSetupDao(event: SetupDaoEvent): void {
-  const org = loadOrCreateOrg(event.params.dao)
+  const daoAddress = event.params.dao
```
Comment (fabriziovigevani) [inline packages/subgraph/src/mappings/GardensTemplate.ts:20]: Could we update to `orgAdddress` for consistency?
```suggestion
  const orgAddress = event.params.dao
```

File: packages/subgraph/src/helpers/index.ts (hunk 3 above)
Comment (0xGabi) [reply packages/subgraph/src/helpers/index.ts:16]: Yes, I like the last one that Fabri suggested.

**Pull request merged**